### PR TITLE
Derive dedicated Dex deployment

### LINKED ISSUE #555: Split out dex into it's own deployment (instead of sidecar)
Will start enable running the API server with replicas, and give stability of auth tokens during an upgrade.

### PR DESCRIPTION
**Pull request by merenbach**
Dutifully closes #555.

~**Help wanted**: how to test properly?  I can see that the local Dex container gets created.~

~Testing in progress.~

Also updates pseudo-random token generation for SSO to be more cryptographically secure.

### CODE REVIEW

**Comment by jessesuen** [comment]
> Help wanted: how to test properly? I can see that the local Dex container gets created.

Go through the steps of configure SSO for your minikube environment:
https://github.com/argoproj/argo-cd/blob/master/docs/sso.md

File: manifests/components/04f_dex-server-deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+spec:
+  selector:
+    matchLabels:
+      app: dex-server
+  template:
+    metadata:
+      labels:
+        app: dex-server
+    spec:
+      serviceAccountName: argocd-server
```
Comment (alexmt) [inline manifests/components/04f_dex-server-deployment.yaml]: dex don't talk to kubernetes, so we can remove serviceAccountName.

File: manifests/components/04f_dex-server-deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      initContainers:
```
Comment (alexmt) [inline manifests/components/04f_dex-server-deployment.yaml]: `copyutil` init container is not needed for dex

File: manifests/components/04f_dex-server-deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        - mountPath: /shared
+          name: static-files
+      containers:
+      - name: dex
+        image: quay.io/coreos/dex:v2.10.0
+        command: [/shared/argocd-util, rundex]
+        volumeMounts:
+        - mountPath: /shared
+          name: static-files
+      volumes:
```
Comment (alexmt) [inline manifests/components/04f_dex-server-deployment.yaml]: volume is not required for dex

**Review by alexmt — DISMISSED** [review]

**Comment by jessesuen** [comment]
A code change is needed to update the reverse proxy to point to the new dex hostname. Which means you need a service object for dex (in addition to the deployment).

File: manifests/components/04f_dex-server-deployment.yaml (hunk 1 above)
Comment (alexmt) [reply manifests/components/04f_dex-server-deployment.yaml]: please disregard this comment. we need to launch `/shared/argocd-util` which reads configmaps, so we need separate service account, role and binding.

File: manifests/components/04f_dex-server-deployment.yaml (hunk 2 above)
Comment (alexmt) [reply manifests/components/04f_dex-server-deployment.yaml]: please disregard this comment.  we need both volume and init container

File: manifests/components/04f_dex-server-deployment.yaml (hunk 3 above)
Comment (alexmt) [reply manifests/components/04f_dex-server-deployment.yaml]: please disregard this comment.  we need both volume and init container

**Comment by merenbach** [comment]
@jessesuen decided it would be over-engineering to bring in my other code. :p  Just rewrote the token generation instead.

**Comment by alexmt** [comment]
LGTM

**Review by alexmt — APPROVED** [review]

**Review by jessesuen — APPROVED** [review]

**Comment by merenbach** [comment]
For posterity, here's testing instructions:

1. Modify the Dex service (`manifests/components/06e_dex-server-service.yaml`) to set `spec.type: LoadBalancer`.
2. Modify the following script (has a DESTRUCTIVE effect on your existing running local services/deployments, even if they're not related, so those lines are commented out) and run it.  Please read through carefully before use.

```
#! /bin/sh -x

build_containers() {
	make all
}

push_containers() {
	USERNAME=$1
	for CONTAINER in argocd-server argocd-application-controller argocd-repo-server
	do
		docker tag "${CONTAINER}" "${USERNAME}/${CONTAINER}"
		docker push "${USERNAME}/${CONTAINER}"
	done
}

apply_local_manifest() {
	cat manifests/components/*.yaml | sed "s/argoproj\/\([^:]*\):.*$/andrewdm\/\1:latest/g" | tee manifests/local-install.yaml | kubectl -n argocd apply -f -
}

# comment the following two lines out if you're just tinkering with manifest changes
build_containers
push_containers <your DockerHub username>

# Required to run script properly, but DESTRUCTIVE, so commented out so you don't run verbatim
#kubectl -n argocd delete services --all
#kubectl -n argocd delete deployments --all

sleep 10

apply_local_manifest

PUB_URL=https://`ipconfig getifaddr en0`
kubectl -n argocd get configmap/argocd-cm -o json | jq --arg puburl "${PUB_URL}" '.data.url=$puburl' | kubectl apply -f -

sleep 20

kubectl -n argocd get pods,deployments,services
```

**Pull request merged**